File: Python DSA/Strings/longest_palindromic_substring.py
"""
LONGEST PALINDROMIC SUBSTRING - Medium
Find the longest palindromic substring in s.

Example: s = "babad" → Output: "bab" or "aba"

Logical Thinking:
1. Expand around center for each position
2. Consider odd length (single center) and even length (two centers)
3. Track maximum palindrome found
4. Alternative: DP but less space efficient

Key: Expand around center, two cases (odd/even)
Time: O(n²) | Space: O(1)
"""

def longest_palindrome(s):
    def expand_around_center(left, right):
        while left >= 0 and right < len(s) and s[left] == s[right]:
            left -= 1
            right += 1
        return right - left - 1
    
    start = end = 0
    for i in range(len(s)):
        len1 = expand_around_center(i, i)
        len2 = expand_around_center(i, i + 1)
        max_len = max(len1, len2)
        
        if max_len > end - start:
            start = i - (max_len - 1) // 2
            end = i + max_len // 2
    
    return s[start:end + 1]


# Test cases
if __name__ == "__main__":
    print(f"Testing Longest Palindromic Substring...")
    print("Add your test cases here")
    print("✅ Refer to problem description for examples")

File: DSA/Strings/Java/LongestPalindromicSubstring.java
// LeetCode 5: Longest Palindromic Substring
// Given a string s, return the longest palindromic substring in s.

public class LongestPalindromicSubstring {
    public String longestPalindrome(String s) {
        if (s == null || s.length() < 1) {
            return "";
        }
        
        int start = 0;
        int end = 0;
        
        for (int i = 0; i < s.length(); i++) {
            int len1 = expandAroundCenter(s, i, i);     // Odd length palindromes
            int len2 = expandAroundCenter(s, i, i + 1); // Even length palindromes
            int len = Math.max(len1, len2);
            
            if (len > end - start) {
                start = i - (len - 1) / 2;
                end = i + len / 2;
            }
        }
        
        return s.substring(start, end + 1);
    }
    
    private int expandAroundCenter(String s, int left, int right) {
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            left--;
            right++;
        }
        return right - left - 1;
    }
    
    // Dynamic Programming approach
    public String longestPalindromeDP(String s) {
        int n = s.length();
        boolean[][] dp = new boolean[n][n];
        String result = "";
        
        // Every single character is a palindrome
        for (int i = 0; i < n; i++) {
            dp[i][i] = true;
            result = s.substring(i, i + 1);
        }
        
        // Check for palindromes of length 2
        for (int i = 0; i < n - 1; i++) {
            if (s.charAt(i) == s.charAt(i + 1)) {
                dp[i][i + 1] = true;
                result = s.substring(i, i + 2);
            }
        }
        
        // Check for palindromes of length 3 and more
        for (int len = 3; len <= n; len++) {
            for (int i = 0; i < n - len + 1; i++) {
                int j = i + len - 1;
                if (s.charAt(i) == s.charAt(j) && dp[i + 1][j - 1]) {
                    dp[i][j] = true;
                    result = s.substring(i, j + 1);
                }
            }
        }
        
        return result;
    }
    
    public static void main(String[] args) {
        LongestPalindromicSubstring solution = new LongestPalindromicSubstring();
        
        String[] testCases = {"babad", "cbbd", "a", "ac"};
        
        for (String test : testCases) {
            System.out.println("\"" + test + "\" -> \"" + solution.longestPalindrome(test) + "\"");
        }
    }
}